dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS/Node.pm

```diff
@@ -64,1412 +276,4508 @@ sub new {
-      # fail with OUT, which is supposed to be updating parameter SVs, not
-      # pushing anything on the stack. So forbid all except RETVAL.
-      $pxs->blurt("Can't use typemap containing DO_ARRAY_ELEM for "
-                    . (defined $out_num ? "OUTLIST" : "OUT")
-                    . " parameter");
-      return;
-    }
-
-    my $subtypemap = $typemaps->get_typemap(ctype => $subtype);
-    if (not $subtypemap) {
-      $pxs->report_typemap_failure($typemaps, $subtype);
-      return;
-    }
-
-    my $suboutputmap = $typemaps->get_outputmap(xstype => $subtypemap->xstype);
-    if (not $suboutputmap) {
-      $pxs->blurt("Error: No OUTPUT definition for type '$subtype', typekind '" . $subtypemap->xstype . "' found");
-      return;
-    }
-
-    my $subexpr = $suboutputmap->cleaned_code;
-    $subexpr =~ s/ntype/subtype/g;
-    $subexpr =~ s/\$arg/ST(ix_$var)/g;
-    $subexpr =~ s/\$var/${var}\[ix_$var]/g;
-    $subexpr =~ s/\n\t/\n\t\t/g;
-    $expr =~ s/\bDO_ARRAY_ELEM\b/$subexpr/;
-
-    # We do our own code emitting and return here (rather than control
-    # passing on to normal RETVAL processing) since that processing is
-    # expecting to push a single temp onto the stack, while our code
-    # pushes several temps.
-    print $pxs->eval_output_typemap_code("qq\a$expr\a", $eval_vars);
-    return;
-  }
-
-
-  # ------------------------------------------------------------------
-  # Now emit code for the three types of return value:
-  #
-  #   RETVAL           - The usual case: store an SV at ST(0) which is set
-  #                      to the value of RETVAL. This is typically a new
-  #                      mortal, but may be optimised to use TARG.
-  #
-  #   OUTLIST param    - if $out_num is defined (and will be >= 0) Push
-  #                      after any RETVAL, new mortal(s) containing the
-  #                      current values of the local var set from that
-  #                      parameter. (May also use TARG if not already used
-  #                      by RETVAL).
-  #
-  #   OUT/OUTPUT param - update passed arg SV at ST($num-1) (which
-  #                      corresponds to param) with the current value of
-  #                      the local var set from that parameter.
-
-  if ($var ne 'RETVAL' and not defined $out_num) {
-    # This is a normal OUTPUT var: i.e. a named parameter whose
-    # corresponding arg on the stack should be updated with the
-    # parameter's current value by using the code contained in the
-    # output typemap.
+    # where everything except $2 and $3 are optional and the 'const'
+    # is for C++ functions.
+
+    my $func_header = shift(@{ $pxs->{line} });
+    $pxs->blurt("Error: cannot parse function definition from '$func_header'"),
+      return
+        unless $func_header =~
+            /^(?:([\w:]*)::)?(\w+)\s*\(\s*(.*?)\s*\)\s*(const)?\s*(;\s*)?$/s;
+
+    my ($class, $name, $params_text) = ($1, $2, $3);
```

Comment:
I realise this was originally `xsub_class`, but given the ways it is used and that it includes the `const` I wonder if the `class` member here would be better named `this_type` or something similar.

It's not entirely true, since THIS is actually `{class} *`, but the "class name" including the `const` seemed strange to me.

Reply:
Perhaps leave the field as called "class", but don't store the "const" string as part of it, and add a separate field called 'const' to indicate whether a 'const' string needs to be emitted?

Reply:
That works for me.